ratex/testing/common.py

```diff
@@ -374,11 +374,58 @@ def run_step(device, model_origin, args, jit_script=True):
     return out
 
 
-def verify_step(model, args, jit_script=True, tol=1e-5):
+def run_step_dx(device, model_origin, args, jit_script=True):
+    """
+    Run the model once.
+
+    Parameters
+    ----------
+    device : str
+        Device to run on
+
+    model_origin : torch.nn.Module
+        The original PyTorch model
+
+    args : List[torch.Tensor]
+        Args of the model
+
+    jit_script :
+        If False, the graph will be traced directly instead of leveraging TorchScript
+        and AutoDiff. This is used to evaluate lowering the ops without backward.
+    """
+    model = copy.deepcopy(model_origin)
+    model.train()
+    model = model.to(device, dtype=torch.float32)
+    if device == "lazy" and jit_script:
+        model = ratex.jit.script(model)
+    args = [arg.to(device) for arg in args]
+    out = model(*args)
+    loss = out.sum()
+    loss.backward()
+    lm.mark_step()
+    return args
+
+
+def verify_step(model, args, jit_script=True, dx=False, tol=1e-5):
     """Verify the results between CPU and Lazy"""
-    torch.testing.assert_close(
-        run_step("cpu", model, args), run_step("lazy", model, args, jit_script), rtol=tol, atol=tol
-    )
+    if dx:
+        args_ratex = []
+        for x in args:
```

Comment:
why we need to have a separate args_ratex ?

Reply:
As you can see, we have 1 reference to args. So in this case, we have 1 tensor that first gets its grad from cpu, but then we then re-forward/auto diff w/ lazy so it gets another grad. This means CPU grad becomes None, so we need to copy. Also, for some reason, lazy requires_grad=True does not work w/ regular torch.randn, so we need to use numpy then create tensor.